Malcore file analyzer #2168

skkra0 opened the pull request:
Closes #2168 

# Description

Created Malcore file analyzer using the executable analysis endpoint

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [x] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [x] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [x] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results. 
    - [x] I have added that raw JSON sample to the `MockUpResponse` of the `_monkeypatch()` method. This serves us to provide a valid sample for testing.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [x] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).

![2024-04-05_19-14](https://github.com/intelowlproject/IntelOwl/assets/73250770/f1021c42-61b5-49d6-bec2-f5b5185d8397)
[report.json](https://github.com/intelowlproject/IntelOwl/files/14891840/report.json)

skkra0 commented:
Hey @mlodic, what tests would you recommend I add for this analyzer?
Also, the Malcore API has another 21 (free to use) endpoints that take files, should I include those as well?

mlodic commented on api_app/analyzers_manager/file_analyzers/malcore_scan.py:
```diff
@@ -0,0 +1,60 @@
+from tests.mock_utils import MockUpResponse, if_mock_connections, patch
+
+
+class MalcoreScan(FileAnalyzer):
+    url = "https://api.malcore.io/api/upload"
+
+    _api_key_name: str
+
+    def run(self):
+        binary = self.read_file_bytes()
+        headers = {"apiKey": self._api_key_name}
+        files = {"filename1": binary}
+
+        try:
+            response = requests.post(self.url, headers=headers, files=files)
+            response.raise_for_status()
+        except requests.RequestException as e:
+            raise AnalyzerRunException(e)
+
+        result = response.json()
+        return result
```
great but actually, as specified in the issue, this should be implemented like other similar sandbox services integrations we have in IntelOwl.
We expect this analyzer to:
* first check whether the file md5/sha256 has already been analyzed recently. In that case, just retrieve the report of the already existing analysis
* if that file is new, upload the file to the service like you just did here
* then, poll an api that checks the status of the analysis, until the analysis is ended
* once the analysis is ended, retrieve the final result

Please check other similar analyzers like [triage](https://github.com/intelowlproject/IntelOwl/blob/master/api_app/analyzers_manager/file_analyzers/triage_scan.py)

mlodic commented on api_app/analyzers_manager/migrations/0075_analyzer_config_malcore_scan.py:
```diff
@@ -0,0 +1,121 @@
+    "name": "Malcore_Scan",
+    "description": "Malcore Scan",
+    "disabled": False,
+    "soft_time_limit": 60,
```
this should be higher and should be higher than the maximum polling time

mlodic commented on api_app/analyzers_manager/migrations/0075_analyzer_config_malcore_scan.py:
```diff
@@ -0,0 +1,121 @@
+    "docker_based": False,
+    "maximum_tlp": "CLEAR",
+    "observable_supported": [],
+    "supported_filetypes": [
+        "application/x-sharedlib",
+        "application/vnd.microsoft.portable-executable",
+        "application/x-elf",
+    ],
```
aren't they supporting more than these type of files?

mlodic commented:
> Hey @mlodic, what tests would you recommend I add for this analyzer?
Also, the Malcore API has another 21 (free to use) endpoints that take files, should I include those as well?

it's enough to populate the monkeypatch like you did.
No need for all the other endpoints. Just implement the ones required for the flow that I have explained in the review

skkra0 commented:
Hey @mlodic,
I rewrote the analyzer, added more file types to the migration file, and raised the time limit.
Can you clarify what you mean by checking whether the file has already been analyzed recently? It seems like IntelOwl already prevents users from repeating scans on a file. I tried renaming a file and scanning it again, but the app still took me back to the previous scan.

mlodic commented on api_app/analyzers_manager/file_analyzers/malcore_scan.py:
```diff
@@ -0,0 +1,333 @@
+
+        binary = self.read_file_bytes()
+        files = {"filename1": binary}
+        try:
+            logger.info(f"Sending {self.md5} to Malcore")
+            response = requests.post(
+                self.base_url + "upload", headers=self.headers, files=files
+            )
+            response.raise_for_status()
+        except requests.RequestException as e:
+            raise AnalyzerRunException(e)
```
if you raise an `AnalyzerRunException` without changing the log name, that would be already handled automatically by the framework. Uncaught errors always trigger AnalyzerRunException

mlodic commented on api_app/analyzers_manager/file_analyzers/malcore_scan.py:
```diff
@@ -0,0 +1,333 @@
+        except TypeError:
+            raise AnalyzerRunException(f"Failed analysis for {self.md5}")
+
+        for _try in range(self.max_tries):
+            logger.info(f"polling malcore try #{_try + 1}")
+            self.result = self._get_status(uuid)["data"]
+            if "status" not in self.result and "msg" not in self.result:
+                break
+
+            time.sleep(self.poll_distance)
```
max_tries and poll_distance must be defined both as a class attribute and in the migration file as params.

mlodic commented on api_app/analyzers_manager/file_analyzers/malcore_scan.py:
```diff
@@ -0,0 +1,333 @@
+            logger.info(f"polling malcore try #{_try + 1}")
+            self.result = self._get_status(uuid)["data"]
+            if "status" not in self.result and "msg" not in self.result:
+                break
```
is this an error? In case, we need to log something about

mlodic replied:
we also needs to log when the final result has been correctly retrieved

mlodic commented on api_app/analyzers_manager/migrations/0075_analyzer_config_malcore_scan.py:
```diff
@@ -0,0 +1,169 @@
+    "docker_based": False,
+    "maximum_tlp": "CLEAR",
+    "observable_supported": [],
+    "supported_filetypes": [
+        "application/x-sharedlib",
+        "application/vnd.microsoft.portable-executable",
+        "application/x-elf",
+        "application/octet-stream",
+        "application/x-binary",
+    ],
```
I would just remove this and keep it empty.

mlodic commented:
> Can you clarify what you mean by checking whether the file has already been analyzed recently? It seems like IntelOwl already prevents users from repeating scans on a file. I tried renaming a file and scanning it again, but the app still took me back to the previous scan

Yep, the default is that the IntelOwl analysis won't be repeated if done in the last 24 hours. see:
![image](https://github.com/intelowlproject/IntelOwl/assets/30625432/c7a76f90-9e1c-4873-a1c6-65b31131870d)

This cache can be overwritten or forced. In that case, you would re-analyze the file again.

However here my point is different.
In case the external service, which is Malcore in this case, already has a public analysis available of that sample, we should avoid to force a new analysis on the same service. So I would check whether the md5 of the file that we upload has already been analyzed in that service. If yes, just retrieve the result of the already available analysis on the external service.
(as already said, in other similar analyzers the same logic has been already implemented...)

skkra0 commented:
> So I would check whether the md5 of the file that we upload has already been analyzed in that service. If yes, just retrieve the result of the already available analysis on the external service.

Hello, sorry for being away for a while. 
Checking if the uploaded file has already been analyzed is a premium feature. Should I still implement it?

mlodic commented:
> Hello, sorry for being away for a while.
Checking if the uploaded file has already been analyzed is a premium feature. Should I still implement it?

no need so

skkra0 commented:
> no need so
no need to check for a public analysis?

mlodic commented:
closing cause not working as expected
